clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -744,6 +750,21 @@ static void addSanitizers(const Triple &TargetTriple,
     // LastEP does not need GlobalsAA.
     PB.registerOptimizerLastEPCallback(SanitizersCallback);
   }
+
+  if (ClRemoveTraps) {
+    // We can optimize after inliner, and PGO profile matching. The hook below
+    // is called from `buildModuleOptimizationPipeline` just after profile use,
+    // and inliner is a part of `buildModuleSimplificationPipeline`, which is
+    // before `buildModuleOptimizationPipeline`.
+    PB.registerOptimizerEarlyEPCallback([&](ModulePassManager &MPM,
```

Comment:
I think this can go at the end of the function simplification pipeline (and there's already a SimplifyCFG run right after) with `registerScalarOptimizerLateEPCallback`. The function should already have been mostly simplified at that point.

Reply:
We need profile matched, which is after simplification

Reply:
profile matching happens right before the inliner/function simplification pipeline (after the initial `EarlyFPM` and `GlobalCleanupPM` cleanup)

Reply:
registerScalarOptimizerLateEPCallback is 22% slower on MultiSource/Benchmark
I'll try to bisect to tell which part is missing.

Reply:
It's bisected to SimplifyCFG , the one after ScalarOptimizerLateEPCallback.
Most trap blocks preceded by empty blocks with unconditional branch as terminator. And they are always cold.
Inserting `SimplifyCFGPass` solves the issue and now I see no regression.

Reply:
why isn't an earlier SimplifyCFG run in the function simplification pipeline handling these?